eclipse-collections-api/src/main/java/org/eclipse/collections/api/ordered/OrderedIterable.java

```diff
@@ -325,6 +326,26 @@
     @Override
     <S, R extends Collection<Pair<T, S>>> R zip(Iterable<S> that, R target);
 
+    /**
+     * Returns a {@code OrderedIterable} formed from this {@code OrderedIterable} and another {@code Iterable} by
+     * applying the {@code Function2}. The second {@code Iterable} should also be ordered.
+     * If one of the two {@code Iterable}s is longer than the other, its
+     * remaining elements are ignored.
+     *
+     * @param that The {@code Iterable} providing the second half of each result pair
+     * @param <S>  The type of the second half of the returned pairs
+     * @param function The function to be applied to each iteration
+     * @return A new {@code OrderedIterable} containing the result of applying the function on each corresponding
+     * element of this {@code OrderedIterable} and that. The length of the returned {@code OrderedIterable} is the
+     * minimum of the lengths of this {@code OrderedIterable} and that.
+     */
+    <S, Z> OrderedIterable<Z> zipWith(Iterable<S> that, Function2<T, S, Z> function);
```

Comment:
The functions should all use bounded wildcards to allow the types to vary more. Specifically, `Function2<T, S, Z>` should be replaced with `Function2<? super T, ? super S, ? extends Z>` in all locations.

Another thing to consider about the api is whether to take Iterable or OrderedIterable. The reason I deprecated zip on RichIterable is that you can write set.zip(otherSet) and you get what feels like random pairs. But if you zip two ordered iterables, you'll get back pairs in a meaningful order. Those methods are deprecated but unchanged for backwards compatibility. On new methods, we can be more strict and change the type of `that` to `OrderedIterable`. This would be consistent with the newish method corresponds(OrderedIterable, Predicate2). We would not need to vary that type on overrides. What do you think?